dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/EventServiceContextBuilder.java

```diff
@@ -0,0 +1,125 @@
+
+    public EventServiceContextBuilder( TrackedEntityInstanceService entityInstanceService,
+        ProgramService programService, TrackedEntityProgramOwnerService trackedEntityProgramOwnerService,
+        OrganisationUnitService organisationUnitService )
+    {
+        this.entityInstanceService = entityInstanceService;
+        this.programService = programService;
+        this.trackedEntityProgramOwnerService = trackedEntityProgramOwnerService;
+        this.organisationUnitService = organisationUnitService;
+    }
+
+    public EventContext build( List<EventRow> eventRowList, User user )
+    {
+        Map<String, List<EventRow>> eventsByProgramUid = eventRowList.stream()
+            .collect( Collectors.groupingBy( EventRow::getProgram ) );
+
+        List<String> trackedEntityInstanceUids = eventRowList.stream()
+            .map( EventRow::getTrackedEntityInstance )
+            .filter( Objects::nonNull )
+            .distinct()
+            .collect( Collectors.toList() );
+
+        Map<String, TrackedEntityInstance> trackedEntityInstanceByUid = entityInstanceService
+            .getTrackedEntityInstancesByUid( trackedEntityInstanceUids, user )
+            .stream()
+            .collect( Collectors.toMap(
+                BaseIdentifiableObject::getUid,
+                trackedEntityInstance -> trackedEntityInstance ) );
+
+        Map<String, Program> programsByUid = programService.getPrograms(
+            eventRowList.stream()
+                .map( EventRow::getProgram )
+                .collect( Collectors.toSet() ) )
+            .stream()
+            .collect( Collectors.toMap(
+                BaseIdentifiableObject::getUid,
+                program -> program ) );
+
+        Map<Pair<String, String>, String> orgUnitByTeiUidAndProgramUidPairs = eventsByProgramUid.keySet().stream()
+            .flatMap( programUid -> trackedEntityProgramOwnerService.getTrackedEntityProgramOwnersUidsUsingId(
+                eventsByProgramUid.get( programUid ).stream()
+                    .map( EventRow::getTrackedEntityInstance )
+                    .filter( Objects::nonNull )
+                    .map( trackedEntityInstanceByUid::get )
+                    .map( BaseIdentifiableObject::getId )
+                    .distinct()
+                    .collect( Collectors.toList() ),
+                programsByUid.get( programUid ) ).stream() )
+            .collect( Collectors.toMap(
+                trackedEntityProgramOwnerIds -> Pair.of(
+                    trackedEntityProgramOwnerIds.getTrackedEntityInstanceId(),
+                    trackedEntityProgramOwnerIds.getProgramId() ),
+                TrackedEntityProgramOwnerIds::getOrgUnitUid ) );
+
+        Map<String, OrganisationUnit> orgUnitsByUid = organisationUnitService
+            .getOrganisationUnitsByUid( orgUnitByTeiUidAndProgramUidPairs.values() ).stream()
+            .collect( Collectors.toMap( BaseIdentifiableObject::getUid, ou -> ou ) );
```

Comment:
@larshelge @bobjolliffe once we retrieved all needed orgUnits uids in previous step, we load them in a single query